### PR TITLE
INFOPLAT-1559 Adds auth header token expiry functionality

### PR DESCRIPTION
**Pull request by hendoxc**
# What
- Adds `BuildAuthHeadersV2` for signing publickey + timestamp.

# Why
- First step towards token expiration using CSA Auth.

# Notes
Further implementation for implementing token refresh will in follow up tickets

### CODE REVIEW

File: pkg/beholder/auth.go [hunk 1]
```diff
@@ -11,17 +13,81 @@ var authHeaderKey = "X-Beholder-Node-Auth-Token"
...
 //
 // where the byte value of <public_key_hex> is what's being signed
-func BuildAuthHeaders(privKey ed25519.PrivateKey) map[string]string {
+// and <signature_hex> is the signature of the public key.
+// The version `2` is:
+//
+// <version>:<public_key_hex>:<timestamp>:<signature_hex>
+//
+// where the byte value of <public_key_hex> and <timestamp> are what's being signed
+func BuildAuthHeaders(privKey ed25519.PrivateKey, opts ...BuildAuthHeadersOpt) map[string]string {
```
Comment (jmank88) [inline pkg/beholder/auth.go:37]: Is there a compelling reason to use functional options? I am worried about how they pollute the package API namespace with verbose functions. Would something simple and direct suffice?
```suggestion
func (AuthHeadersConfig) Build(privKey ed25519.PrivateKey) map[string]string {
```

Comment (hendoxc) [reply pkg/beholder/auth.go:37]: Allows existing backwards compatibility of `BuildAuthHeaders`. But I agree,  I think the most direct approach here would be to expose `BuildAuthHeadersV2`

Comment (jmank88) [reply pkg/beholder/auth.go:37]: `BuildAuthHeadersV2` sounds good, but FWIW adding variadic args may not break callers, but it does change the function signature and is not compatible by definition. This article is a good read: https://go.dev/blog/module-compatibility#adding-to-a-function
> Sometimes the techniques of adding a new function and adding options can be combined by making the options struct a method receiver. Consider the evolution of the net package’s ability to listen at a network address.

https://pkg.go.dev/net@go1.11#ListenConfig

Comment (hendoxc) [reply pkg/beholder/auth.go]: Thanks ^ this was helpful, I took a look at the `crypto/tls` `Dial` example

File: pkg/beholder/auth.go [hunk 2]
```diff
@@ -2,26 +2,75 @@ package beholder
...
 	"fmt"
+	"time"
 )
 
 // authHeaderKey is the name of the header that the node authenticator will use to send the auth token
 var authHeaderKey = "X-Beholder-Node-Auth-Token"
 
 // authHeaderVersion is the version of the auth header format
 var authHeaderVersion = "1"
+var authHeaderVersionV2 = "2"
```
Comment (jmank88) [inline pkg/beholder/auth.go:15]: nit/ VersionVersion? :shrug: 
```suggestion
var authHeaderVersion2 = "2"
```

**Review by jmank88 — APPROVED** [review]

File: pkg/beholder/auth.go [hunk 3]
```diff
@@ -2,26 +2,75 @@ package beholder
...
+		config = defaultAuthHeaderConfig()
+	}
+	if config.version == "" {
+		config.version = authHeaderVersion2
+	}
+	// If timestamp is not set, use the current time
+	if config.timestamp == 0 {
+		config.timestamp = time.Now().UnixMilli()
+	}
+	// If timestamp is negative, set it to 0. negative values cause overflow on convertsion to uint64
```
Comment (4of9) [inline pkg/beholder/auth.go:54]:
```suggestion
	// If timestamp is negative, set it to 0. negative values cause overflow on conversion to uint64
```

File: pkg/beholder/auth.go [hunk 4]
```diff
@@ -2,26 +2,75 @@ package beholder
...
+	// 0 timestamps will be rejected by the server as being too old
+	if config.timestamp < 0 {
+		config.timestamp = 0
```
Comment (4of9) [inline pkg/beholder/auth.go:57]: this sorta conflicts in spirit to the previous block where we set the timestamp to `now` if it's `0`

should this be moved up?

File: pkg/beholder/auth_test.go [hunk 5]
```diff
@@ -2,21 +2,137 @@ package beholder
...
+	assert.Equal(t, csaPrivKey.Public().(ed25519.PublicKey), ed25519.PublicKey(pubKeyBytes))
+
+	// Parse the timestamp
+	timestampParsed, err := strconv.ParseInt(timestampStr, 10, 64)
+	require.NoError(t, err)
+
+	// Verify the timestamp is within the last 5 seconds
+	// This verifies that default configuration is to use the current time
+	now := time.Now().UnixMilli()
+	assert.InDelta(t, now, timestampParsed, 5000, "timestamp should be within the last 5 seconds")
```
Comment (4of9) [inline pkg/beholder/auth_test.go:96]: nitpick: could tighten up the time range by recording `now` at the start of the test, sleeping a few milliseconds, and then comparing against that

File: pkg/beholder/auth.go (hunk 4 above)
Comment (hendoxc) [reply pkg/beholder/auth.go:57]: So you're saying if a negative value is passed in, we should auto correct to current time?

Comment (hendoxc) [reply pkg/beholder/auth.go:57]: I think let's keep as is then

File: pkg/beholder/auth_test.go [hunk 6]
```diff
@@ -2,21 +2,138 @@ package beholder
...
+	require.NoError(t, err)
+	assert.Equal(t, csaPrivKey.Public().(ed25519.PublicKey), ed25519.PublicKey(pubKeyBytes))
+
+	// Parse the timestamp
+	timestampParsed, err := strconv.ParseInt(timestampStr, 10, 64)
+	require.NoError(t, err)
+
+	// Verify the timestamp is within the last 50ms
+	// This verifies that default configuration is to use the current time
+	assert.InDelta(t, now, timestampParsed, 50, "timestamp should be within the last 50ms")
```
Comment (pkcll) [inline pkg/beholder/auth_test.go:97]: nit: Should it be just after or equal `now` ?  or you can use mock clock github.com/jonboulle/clockwork

**Review by pkcll — APPROVED** [review]

File: pkg/beholder/auth.go [hunk 7]
```diff
@@ -2,26 +2,75 @@ package beholder
...
 	pubKey := privKey.Public().(ed25519.PublicKey)
 	messageBytes := pubKey
 	signature := ed25519.Sign(privKey, messageBytes)
-	headerValue := fmt.Sprintf("%s:%x:%x", authHeaderVersion, messageBytes, signature)
 
-	return map[string]string{authHeaderKey: headerValue}
+	return map[string]string{authHeaderKey: fmt.Sprintf("%s:%x:%x", authHeaderVersion1, messageBytes, signature)}
+}
+
+func BuildAuthHeadersV2(privKey ed25519.PrivateKey, config *AuthHeaderConfig) map[string]string {
```
Comment (cll-gg) [inline pkg/beholder/auth.go:43]:
```suggestion
// BuildAuthHeadersV2 creates the auth header value to be included on requests.
// See documentation on BuildAuthHeaders for more info. 
func BuildAuthHeadersV2(privKey ed25519.PrivateKey, config *AuthHeaderConfig) map[string]string {
```

File: pkg/beholder/auth.go [hunk 8]
```diff
@@ -2,26 +2,75 @@ package beholder
...
+	if config == nil {
+		config = defaultAuthHeaderConfig()
+	}
+	if config.version == "" {
+		config.version = authHeaderVersion2
+	}
```
Comment (cll-gg) [inline pkg/beholder/auth.go:49]: Not sure how important it is, but with the current code we theoretically allow a caller to build v2 headers while passing in a different `version`, which will get set in the header key.

File: pkg/beholder/auth.go [hunk 9]
```diff
@@ -2,26 +2,75 @@ package beholder
...
+	messageBytes := append(pubKey, timestampUnixMsBytes...)
+	signature := ed25519.Sign(privKey, messageBytes)
+
+	return map[string]string{authHeaderKey: fmt.Sprintf("%s:%x:%d:%x", config.version, pubKey, config.timestamp, signature)}
+}
+
+func defaultAuthHeaderConfig() *AuthHeaderConfig {
+	return &AuthHeaderConfig{
+		version:   authHeaderVersion2,
+		timestamp: time.Now().UnixMilli(),
```
Comment (cll-gg) [inline pkg/beholder/auth.go:74]: Is it intended that we don't use `config.timestamp` here? Seems counter-intuitive

Comment (hendoxc) [reply pkg/beholder/auth.go:76]: not sure I understand

File: pkg/beholder/auth.go (hunk 8 above)
Comment (hendoxc) [reply pkg/beholder/auth.go:51]: Yes, by design, it means this function can support upgrading to a new version

File: pkg/beholder/auth.go (hunk 9 above)
Comment (cll-gg) [reply pkg/beholder/auth.go:76]: Ah I think I mixed things up, nvm!

**Review by cll-gg — APPROVED** [review]

**Review by 4of9 — APPROVED** [review]

**Comment by hendoxc** [comment]
Closing in favour of
https://github.com/smartcontractkit/chainlink-common/pull/974

**Pull request closed without merging**